psyneulink/core/llvm/helpers.py

```diff
@@ -668,4 +669,105 @@ def generate_sched_condition(self, builder, condition, cond_ptr, node,
+                        depth += 1
+                    else:
+                        break
+
+                return depth
+
+            target = condition.dependency
+            param = condition.parameter
+            threshold = condition.threshold
+            comparator = condition.comparator
+            indices = condition.indices
+
+            tolerance_value = condition.rtol * abs(threshold) + condition.atol
+
+            assert param in target.llvm_state_ids, (
+                f"Threshold for {target} only supports items in llvm_state_ids"
+                f" ({target.llvm_state_ids})"
+            )
+
+            node_state = builder.gep(
+                nodes_states,
+                [
+                    self.ctx.int32_ty(0),
+                    self.ctx.int32_ty(self.composition._get_node_index(target))
+                ]
+            )
+
+            param_ptr = get_state_ptr(builder, target, node_state, param)
+
+            assert isinstance(
+                param_ptr.type.pointee,
+                (ir.ArrayType, ir.IntType, ir.DoubleType, ir.FloatType)
+            ), f"Unsupported type of '{param}' ({param_ptr.type})"
+
+            if indices is None:
+                single_elem_depth = get_single_element_depth(param_ptr)
+                assert single_elem_depth is not None, f"'{param}' contains multiple numbers but no indices were specified"
+
+                # get single number from array of any dimension
+                val = builder.load(
+                    builder.gep(
+                        param_ptr,
+                        [self.ctx.int32_ty(0) for _ in range(single_elem_depth + 1)]
+                    )
+                )
+            else:
+                if isinstance(indices, TimeScale):
+                    indices = [indices.value]
+
+                val = builder.load(
+                    builder.gep(
+                        param_ptr,
+                        [self.ctx.int32_ty(x) for x in ([0] + list(indices))]
+                    )
+                )
+
+            if is_integer(val) and isinstance(threshold, int) and isinstance(tolerance_value, int):
+                cmp_method = builder.icmp_signed
+                abs_method_name = 'abs'
+            elif is_floating_point(val) or isinstance(threshold, float) or isinstance(tolerance_value, float):
+                val = convert_type(builder, val, self.ctx.float_ty)
+                cmp_method = builder.fcmp_ordered
+                abs_method_name = 'fabs'
+            else:
+                assert False, f"Unsupported type of '{param}' ({val.type}) or threshold ({type(threshold)})"
+
+            threshold = val.type(threshold)
+
+            if tolerance_value == 0:
+                return cmp_method(comparator, val, threshold)
+            else:
+                abs_method = self.ctx.get_builtin(abs_method_name, [val.type])
+                val_thresh_diff = builder.call(abs_method, [builder.fsub(val, threshold)])
```

Comment:
True, is it better to handle int/float separately instead of just converting int to float? If any/all of the comparisons are to be done with helpers.is_close, I think that just takes floats.

Reply:
I was thinking about it too. with the current default of fp64 it should be ok (all int32 values can be represented in fp64).
Switching to fp32 will introduce problems for values above 16.7M, which might be too low for lifetime execution count.
I think it'd make sense to special case execution counts as the only  integer codepath and maybe restrict allowed tolerances (e.g. only integer absolute tolerance?)
similarly to above, I don't have a good idea about what the users would expect here.
Conversion to fp64 and using is_close should be safe, for all situations, but might be unnecessary.

Reply:
I'm fine with writing a special case for execution count if it's faster or otherwise better, but for now I've just converted